File: output/cpp/tests/36_context_basic.cpp
#include <string>
#include <unordered_map>
#include <vector>
#include <any>
#include <memory>


#include <iostream>
#include <string>
#include <cassert>

// Test: Basic System Context Access
// Validates @@.param, @@:return, @@:event syntax

class ContextBasicTestFrameEvent {
public:
    std::string _message;
    std::unordered_map<std::string, std::any> _parameters;

    ContextBasicTestFrameEvent(const std::string& message, std::unordered_map<std::string, std::any> params = {})
        : _message(message), _parameters(std::move(params)) {}
};

class ContextBasicTestFrameContext {
public:
    ContextBasicTestFrameEvent _event;
    std::any _return;
    std::unordered_map<std::string, std::any> _data;

    ContextBasicTestFrameContext(ContextBasicTestFrameEvent event, std::any default_return = {})
        : _event(std::move(event)), _return(std::move(default_return)) {}
};

class ContextBasicTestCompartment {
public:
    std::string state;
    std::unordered_map<std::string, std::any> state_args;
    std::unordered_map<std::string, std::any> state_vars;
    std::unordered_map<std::string, std::any> enter_args;
    std::unordered_map<std::string, std::any> exit_args;
    std::unique_ptr<ContextBasicTestFrameEvent> forward_event;
    std::unique_ptr<ContextBasicTestCompartment> parent_compartment;

    explicit ContextBasicTestCompartment(const std::string& state) : state(state) {}
};

class ContextBasicTest {
private:
    std::unique_ptr<ContextBasicTestCompartment> __compartment;
    std::unique_ptr<ContextBasicTestCompartment> __next_compartment;
    std::vector<std::unique_ptr<ContextBasicTestCompartment>> _state_stack;
    std::vector<ContextBasicTestFrameContext> _context_stack;

    void __kernel(ContextBasicTestFrameEvent& __e) {
        __router(__e);
        while (__next_compartment) {
            auto next_compartment = std::move(__next_compartment);
            ContextBasicTestFrameEvent exit_event("<$");
            __router(exit_event);
            __compartment = std::move(next_compartment);
            if (!__compartment->forward_event) {
                ContextBasicTestFrameEvent enter_event("$>");
                __router(enter_event);
            } else {
                auto forward_event = std::move(__compartment->forward_event);
                if (forward_event->_message == "$>") {
                    __router(*forward_event);
                } else {
                    ContextBasicTestFrameEvent enter_event("$>");
                    __router(enter_event);
                    __router(*forward_event);
                }
            }
        }
    }

    void __router(ContextBasicTestFrameEvent& __e) {
        const auto& state_name = __compartment->state;
        if (state_name == "Ready") {
            _state_Ready(__e);
        }
    }

    void __transition(std::unique_ptr<ContextBasicTestCompartment> next) {
        __next_compartment = std::move(next);
    }

    void _state_Ready(ContextBasicTestFrameEvent& __e) {
        if (__e._message == "add") {
            auto a = std::any_cast<int>(__e._parameters.at("a"));
            auto b = std::any_cast<int>(__e._parameters.at("b"));
            {
            _context_stack.back()._return = @@.a + @@.b;
            }
            return;
        } else if (__e._message == "get_event_name") {
            {
            _context_stack.back()._return = @@:event;
            }
            return;
        } else if (__e._message == "greet") {
            auto name = std::any_cast<std::string>(__e._parameters.at("name"));
            {
            std::string result = std::string("Hello, ") + @@.name + "!";
            _context_stack.back()._return = result;
            }
            return;
        }
    }

public:
    ContextBasicTest() {
        __compartment = std::make_unique<ContextBasicTestCompartment>("Ready");
        ContextBasicTestFrameEvent __frame_event("$>");
        __kernel(__frame_event);
    }

    int add(int a, int b) {
        std::unordered_map<std::string, std::any> __params;
        __params["a"] = a;
        __params["b"] = b;
        ContextBasicTestFrameEvent __e("add", std::move(__params));
        ContextBasicTestFrameContext __ctx(std::move(__e), std::any(int()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<int>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string get_event_name() {
        ContextBasicTestFrameEvent __e("get_event_name");
        ContextBasicTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

    std::string greet(std::string name) {
        std::unordered_map<std::string, std::any> __params;
        __params["name"] = name;
        ContextBasicTestFrameEvent __e("greet", std::move(__params));
        ContextBasicTestFrameContext __ctx(std::move(__e), std::any(std::string()));
        _context_stack.push_back(std::move(__ctx));
        __kernel(_context_stack.back()._event);
        auto __result = std::any_cast<std::string>(std::move(_context_stack.back()._return));
        _context_stack.pop_back();
        return __result;
    }

};

int main() {
    printf("=== Test 36: Context Basic ===\n");
    ContextBasicTest s;

    int result1 = s.add(3, 5);
    if (result1 != 8) {
        printf("FAIL: Expected 8, got %d\n", result1);
        assert(false);
    }
    printf("1. add(3, 5) = %d\n", result1);

    std::string eventName = s.get_event_name();
    if (eventName != "get_event_name") {
        printf("FAIL: Expected 'get_event_name', got '%s'\n", eventName.c_str());
        assert(false);
    }
    printf("2. @@:event = '%s'\n", eventName.c_str());

    std::string greeting = s.greet("World");
    if (greeting != "Hello, World!") {
        printf("FAIL: Expected 'Hello, World!', got '%s'\n", greeting.c_str());
        assert(false);
    }
    printf("3. greet('World') = '%s'\n", greeting.c_str());

    printf("PASS: Context basic access works correctly\n");
    return 0;
}

File: output/go/tests/36_context_basic.go

package main

import "fmt"
import "os"

type ContextBasicTestFrameEvent struct {
    _message    string
    _parameters map[string]any
}

type ContextBasicTestFrameContext struct {
    _event  ContextBasicTestFrameEvent
    _return any
    _data   map[string]any
}

type ContextBasicTestCompartment struct {
    state            string
    stateArgs        map[string]any
    stateVars        map[string]any
    enterArgs        map[string]any
    exitArgs         map[string]any
    forwardEvent     *ContextBasicTestFrameEvent
    parentCompartment *ContextBasicTestCompartment
}

func newContextBasicTestCompartment(state string) *ContextBasicTestCompartment {
    return &ContextBasicTestCompartment{
        state:    state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
    }
}

func (c *ContextBasicTestCompartment) copy() *ContextBasicTestCompartment {
    nc := &ContextBasicTestCompartment{
        state: c.state,
        stateArgs: make(map[string]any),
        stateVars: make(map[string]any),
        enterArgs: make(map[string]any),
        exitArgs:  make(map[string]any),
        forwardEvent:     c.forwardEvent,
        parentCompartment: c.parentCompartment,
    }
    for k, v := range c.stateArgs { nc.stateArgs[k] = v }
    for k, v := range c.stateVars { nc.stateVars[k] = v }
    for k, v := range c.enterArgs { nc.enterArgs[k] = v }
    for k, v := range c.exitArgs { nc.exitArgs[k] = v }
    return nc
}

type ContextBasicTest struct {
    _state_stack []*ContextBasicTestCompartment
    __compartment *ContextBasicTestCompartment
    __next_compartment *ContextBasicTestCompartment
    _context_stack []ContextBasicTestFrameContext
}

func NewContextBasicTest() *ContextBasicTest {
    s := &ContextBasicTest{}
    s._state_stack = make([]*ContextBasicTestCompartment, 0)
    s._context_stack = make([]ContextBasicTestFrameContext, 0)
    s.__compartment = newContextBasicTestCompartment("Ready")
    s.__next_compartment = nil
    __frame_event := ContextBasicTestFrameEvent{_message: "$>", _parameters: nil}
    __ctx := ContextBasicTestFrameContext{_event: __frame_event, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return s
}

func (s *ContextBasicTest) __kernel(__e *ContextBasicTestFrameEvent) {
    s.__router(__e)
    for s.__next_compartment != nil {
        next_compartment := s.__next_compartment
        s.__next_compartment = nil
        exit_event := &ContextBasicTestFrameEvent{_message: "<$", _parameters: s.__compartment.exitArgs}
        s.__router(exit_event)
        s.__compartment = next_compartment
        if s.__compartment.forwardEvent == nil {
            enter_event := &ContextBasicTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
            s.__router(enter_event)
        } else {
            forward_event := s.__compartment.forwardEvent
            s.__compartment.forwardEvent = nil
            if forward_event._message == "$>" {
                s.__router(forward_event)
            } else {
                enter_event := &ContextBasicTestFrameEvent{_message: "$>", _parameters: s.__compartment.enterArgs}
                s.__router(enter_event)
                s.__router(forward_event)
            }
        }
    }
}

func (s *ContextBasicTest) __router(__e *ContextBasicTestFrameEvent) {
    switch s.__compartment.state {
    case "Ready":
        s._state_Ready(__e)
    }
}

func (s *ContextBasicTest) __transition(next *ContextBasicTestCompartment) {
    s.__next_compartment = next
}

func (s *ContextBasicTest) Add(a int, b int) int {
    __params := map[string]any{
        "a": a,
        "b": b,
    }
    __e := ContextBasicTestFrameEvent{_message: "Add", _parameters: __params}
    __ctx := ContextBasicTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result int
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(int) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextBasicTest) GetEventName() string {
    __e := ContextBasicTestFrameEvent{_message: "GetEventName"}
    __ctx := ContextBasicTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextBasicTest) Greet(name string) string {
    __params := map[string]any{
        "name": name,
    }
    __e := ContextBasicTestFrameEvent{_message: "Greet", _parameters: __params}
    __ctx := ContextBasicTestFrameContext{_event: __e, _data: make(map[string]any)}
    s._context_stack = append(s._context_stack, __ctx)
    s.__kernel(&s._context_stack[len(s._context_stack)-1]._event)
    var __result string
    if __rv := s._context_stack[len(s._context_stack)-1]._return; __rv != nil { __result = __rv.(string) }
    s._context_stack = s._context_stack[:len(s._context_stack)-1]
    return __result
}

func (s *ContextBasicTest) _state_Ready(__e *ContextBasicTestFrameEvent) {
    if __e._message == "Add" {
        a := __e._parameters["a"].(int)
        _ = a
        b := __e._parameters["b"].(int)
        _ = b
        // Access params directly (Go backend extracts them)
        s._context_stack[len(s._context_stack)-1]._return = a + b
    } else if __e._message == "GetEventName" {
        // Access event name via __e
        s._context_stack[len(s._context_stack)-1]._return = __e._message
    } else if __e._message == "Greet" {
        name := __e._parameters["name"].(string)
        _ = name
        // Mix param access and return
        result := "Hello, " + name + "!"
        s._context_stack[len(s._context_stack)-1]._return = result
    }
}

func main() {
	fmt.Println("=== Test 36: Context Basic ===")
	sm := NewContextBasicTest()

	// Test 1: param access, @@:return
	result1 := sm.Add(3, 5)
	if result1 != 8 {
		fmt.Printf("FAIL: Expected 8, got %d\n", result1)
		os.Exit(1)
	}
	fmt.Printf("1. Add(3, 5) = %d\n", result1)

	// Test 2: event name access
	eventName := sm.GetEventName()
	if eventName != "GetEventName" {
		fmt.Printf("FAIL: Expected 'GetEventName', got '%s'\n", eventName)
		os.Exit(1)
	}
	fmt.Printf("2. event = '%s'\n", eventName)

	// Test 3: param access with string
	greeting := sm.Greet("World")
	if greeting != "Hello, World!" {
		fmt.Printf("FAIL: Expected 'Hello, World!', got '%s'\n", greeting)
		os.Exit(1)
	}
	fmt.Printf("3. Greet('World') = '%s'\n", greeting)

	fmt.Println("PASS: Context basic access works correctly")
}
